### ISSUE TITLE
Suffixing key::Tab to $value in Selenium2Driver::setValue() scroll page in some cases

### ISSUE THREAD

**Issue by rajeshtaneja** [minkphp/MinkSelenium2Driver#194]
With key::Tab suffixed to $value page is scrolled in some cases and current elements are not intractable

Offset within element cannot be scrolled into view: (15, 40.5): [object HTMLDivElement]
      Build info: version: '2.44.0', revision: '76d78cf', time: '2014-10-23 20:02:37'  
      System info: host: 'rajesh', ip: '127.0.1.1', os.name: 'Linux', os.arch: 'amd64', os.version: '3.16.0-25-generic', java.version: '1.7.0_65'  
      Driver info: driver.version: unknown (WebDriver\Exception\MoveTargetOutOfBounds)

**Comment by aik099** [comment]
Can you provide a test case for this? Is this scrolling related to specific page content or JavaScript code present there?

**Comment by rajeshtaneja** [comment]
Sorry for the late response Alexander.

Text fields are being created by JS and it's happening in 1 specific case for now.
I will try create a specific testcase for this, but for now if you can install moodle you can see problem at https://github.com/moodle/moodle/blob/master/grade/grading/form/rubric/tests/behat/edit_rubric.feature#L32

**Comment by aik099** [comment]
Not very standalone test case. Can you please create test case with Mink only calls, that doesn't involve installing any frameworks or Behat?

**Comment by rajeshtaneja** [comment]
I will try make a standalone test-case soon.

**Comment by 4rthem** [comment]
I also need to set a value without the TAB key.
The [select2](https://github.com/select2/select2) listens to the TAB key to select the first element of the list.

IMO this also could be usefull to avoid triggering the `change` event for other reasons.

```php
Behat\Mink\Driver\Selenium2Driver::setValue($xpath, $value, $appendTab = true);
```
or

```php
Behat\Mink\Driver\Selenium2Driver::setValue($xpath, $value, $triggerChange = true);
```

**Comment by aik099** [comment]
We trigger `change` event in same cases, when browser does.

Any other solution how to avoid double `change` event being triggered and without using the `TAB`?

The modification of `setValue` method will make all other driver inconsistent and looks more like selenium2 hack.

**Comment by aik099** [comment]
We have tests in place to check that `change` event is fired, when it should and only once. So you can safely alter `setValue` to get desired result as long as tests pass.

**Comment by aik099** [comment]
Another issue, where TAB sending causes issue: #188

**Comment by aronbeal** [comment]
I've run into this issue as well.  There is no option given to exclude the tab key from the posted value, as it is presumed to be necessary to trigger the change event.  In my work with Drupal 7's autocomplete field, however, the field reacts to the onChange event, and the tab will submit the value.  In fields where the user must choose one of the predefined values, this causes an exception.

Perhaps something akin to the following (for Selenium2Driver.php setValue method), in order to ensure backward compatibility?:

```php
**
 * {@inheritdoc}
 */
public function setValue($xpath, $value, $options=array()) {
  $options = $options + array(
    'addTab'=>TRUE
  );
  $element = $this->findElement($xpath);
  $elementName = strtolower($element->name());

  if ('select' === $elementName) {
    if (is_array($value)) {
      $this->deselectAllOptions($element);

      foreach ($value as $option) {
        $this->selectOptionOnElement($element, $option, TRUE);
      }

      return;
    }

    $this->selectOptionOnElement($element, $value);

    return;
  }

  if ('input' === $elementName) {
    $elementType = strtolower($element->attribute('type'));

    if (in_array($elementType, array('submit', 'image', 'button', 'reset'))) {
      throw new DriverException(sprintf('Impossible to set value an element with XPath "%s" as it is not a select, textarea or textbox', $xpath));
    }

    if ('checkbox' === $elementType) {
      if ($element->selected() xor (bool) $value) {
        $this->clickOnElement($element);
      }

      return;
    }

    if ('radio' === $elementType) {
      $this->selectRadioValue($element, $value);

      return;
    }

    if ('file' === $elementType) {
      $element->postValue(array('value' => array(strval($value))));

      return;
    }
  }

  $value = strval($value);

  if (in_array($elementName, array('input', 'textarea'))) {
    $existingValueLength = strlen($element->attribute('value'));
    // Add the TAB key to ensure we unfocus the field as browsers are triggering the change event only
    // after leaving the field.
    $value = str_repeat(Key::BACKSPACE . Key::DELETE, $existingValueLength) . $value . (($options['addTab']) ? Key::TAB : '');
  }

  $element->postValue(array('value' => array($value)));
}
```

A diff of the file from the original with these changes would be:

```bash
290d289
< 
636d634
<   public function setValue($xpath, $value) {
637a636,639
>   public function setValue($xpath, $value, $options=array()) {
>     $options = $options + array(
>       'addTab'=>TRUE
>     );
691c693
<       $value = str_repeat(Key::BACKSPACE . Key::DELETE, $existingValueLength) . $value . Key::TAB;
---
>       $value = str_repeat(Key::BACKSPACE . Key::DELETE, $existingValueLength) . $value . (($options['addTab']) ? Key::TAB : '');
```

**Comment by aik099** [comment]
Fact, that this driver only internally uses TAB to avoid duplicate `change` event problem is the way how it does it. Allowing to change that behavior creates ability for user to alter internal behavior of the driver as they see fit and that's no good.

Alternatively any implementation of double `change` event protection can be implemented which doesn't use tab to allow working with autocomplete fields.

But value change on a field will trigger a `change` event not matter if driver sends TAB internally or not. So I don't see how not sending TAB would help.

**Comment by aronbeal** [comment]
In my case, it's not that I'm trying to suppress a change event, but rather the loss of field focus.

What happens is the field comes back with an autocomplete list after a (buffered) text change event on the field.  This is good.  When the field loses focus (via tab), it's counted as an ajax submission, and the system complains because the sent value is not one of the values it knows about (in particular, when the user chooses a value, a number is parenthetically appended to the value which the system looks for in any valid submission).  This is the bad part.

AFAIK this is stock Drupal 7 behavior, not something that we've ginned up special.

**Comment by aronbeal** [comment]
So, I was able to work around this with the following:

```php
    //assume class this method snippet exists within extends 
    //Drupal\DrupalExtension\Context\RawDrupalContext,
    //field_text contains the string I want to enter in the field,
    //and element implements Behat\Gherkin\Node\NodeInterface
    $session = $this->getSession();
    $xpath = $element->getXpath();
    $driver = $session->getDriver();
    $chars = str_split($field_text);
    $driver->focus($xpath);

    foreach($chars as $char){
      $driver->keyDown($xpath, $char);
      $driver->keyUp($xpath, $char);
    }
```

This evokes the desired behavior without the appended tab key.  This will solve my issues without any modification of setValue

**Comment by aik099** [comment]
I absolutely agree that current implementation with tab does break autocomplete controls from different vendors. I just don't know what other implementation could be created to keep all driver tests passing and using the tab.

**Comment by aronbeal** [comment]
Would not what I posted above work?  The third parameter, $options, is provided with defaults, so it's not required.  From an external perspective, the function signature looks the same, and the default behavior is to keep the tab, so the behavior is the same as well.

**Comment by aronbeal** [comment]
I see you're trying to adhere to both letter and spirit of DriverInterface.php.  Alternative suggestion: make findElement public?  Then people who need it can just implement their own setValue method external to SeleniumDriver that excludes the tab.  It's not part of the interface, so this seems possible?  Albeit kludgy?

**Comment by aik099** [comment]
The `NodeElement` already allows searching for elements. No need to build xpath by hand.

**Comment by aronbeal** [comment]
I spoke too soon.  I could have sworn I saw text being entered appropriately using this approach, but I can no longer recreate it.  I tried switching to entering via element like you suggested:

```php
    $element->focus();
    for($i = 65; $i <= 90; $i++){
      $element->keyDown($i);
      $element->keyUp($i);
    }
```
This *should* be printing A-Z in the field, but it's not. It's focusing correctly, but nothing is being output to the field.

**Comment by aik099** [comment]
In your approach the keydown/keyup events are triggered for each letter typed obviously.

**Comment by aik099** [comment]
Should be solved once #244 is merged.

**Comment by stof** [comment]
closing as the PR has been merged. Please reopen the issue if the PR have not actually fixed it.
